### PR TITLE
[Snyk] Security upgrade keras from 2.4.3 to 3.13.2

### PR DESCRIPTION
**Pull request by rwils83**
snyk-top-banner

#### Snyk changed the following file(s):

- `requirements.txt`

<details>
<summary>⚠️ <b>Warning</b></summary>

```
terminado 0.10.1 requires tornado, which is not installed.
notebook 6.4.0 requires tornado, which is not installed.
nbconvert 6.3.0b0 has requirement mistune<2,>=0.8.1, but you have mistune 2.0.3.
jupyter-client 6.1.12 requires tornado, which is not installed.
ipykernel 6.0.1 requires tornado, which is not installed.
ipykernel 6.0.1 has requirement importlib-metadata<4; python_version < "3.8.0", but you have importlib-metadata 6.7.0.
```

</details>

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/purplehax/project/18948368-1688-4204-8c8a-e0f424001472?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/purplehax/project/18948368-1688-4204-8c8a-e0f424001472?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Learn about vulnerability in an interactive lesson of Snyk Learn.](https://learn.snyk.io/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":false,"FF_showPullRequestBreakingChangesWebSearch":false,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"keras","from":"2.4.3","to":"3.13.2"}],"env":"prod","issuesToFix":["SNYK-PYTHON-KERAS-15268069"],"prId":"205f3645-7e01-4af6-a2d0-22cdbd264893","prPublicId":"205f3645-7e01-4af6-a2d0-22cdbd264893","packageManager":"pip","priorityScoreList":[641],"projectPublicId":"18948368-1688-4204-8c8a-e0f424001472","projectUrl":"https://app.snyk.io/org/purplehax/project/18948368-1688-4204-8c8a-e0f424001472?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","pr-warning-shown","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-KERAS-15268069"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by rwils83** [comment]
Merge Risk: High

The upgrade from Keras 2.4.3 to 3.13.2 is a major rewrite with significant breaking changes. Keras 3 transitions from being a TensorFlow-specific high-level API (`tf.keras`) to a multi-backend framework that can run on top of TensorFlow, JAX, or PyTorch. [1, 7]  
  
**Key Breaking Changes:**  
  
*   **Package and Import Paths:** This is the most critical change. All imports must be updated across your project. For example, `from tensorflow.keras import layers` must be changed to `from keras import layers`. [3, 8]  
*   **Multi-Backend System:** Keras 3 requires you to choose and configure a backend (TensorFlow, JAX, or PyTorch). [1] While existing code might run with the TensorFlow backend, any direct use of TensorFlow-specific APIs (e.g., `tf.linalg`) will break portability and may require refactoring to use the new backend-agnostic `keras.ops` module. [3]  
*   **Python Version Requirement:** Starting with version 3.13.0, Keras requires Python 3.11 or higher. [6]  
*   **Model Saving:** Saving models to the TensorFlow SavedModel format via `model.save()` is no longer supported. The recommended format is the new `.keras` v3 format. [5, 8]  
  
**Recommendation:** This is a major migration that requires significant code refactoring, dependency management (choosing a backend), and thorough testing. It should be handled as a dedicated engineering effort. For projects that are not actively developed but need to run on updated dependencies, an alternative is to switch to the `tf-keras` package, which maintains the Keras 2 API. [1, 10]  
  
Source: [Keras 3 Migration Guide](https://keras.io/guides/migrating_to_keras_3/)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.
